secrets/store_test.go

```diff
@@ -95,6 +96,61 @@ func TestNewStore(t *testing.T) {
 	}
 }
 
+func TestNewStoreMiddleware(t *testing.T) {
+	dir, err := ioutil.TempDir("", "secret_test_")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tmpFile, err := ioutil.TempFile(dir, "secrets.json")
+	if err != nil {
+		t.Fatal(err)
+	}
+	tmpFile.Write([]byte(specificationExample))
+
+	store, err := NewStore(context.Background(), tmpFile.Name(), log.TestWrapper(t))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var (
+		expectedMiddlewareCalls = 2
+		middlewareCall          int
+		wg                      sync.WaitGroup
+	)
+
+	wg.Add(expectedMiddlewareCalls)
+	var middleware = func(next SecretHandlerFunc) SecretHandlerFunc {
+		return func(sec *Secrets) {
+			middlewareCall++
+			wg.Done()
+			next(sec)
+		}
+	}
+	store.SecretHandler(middleware, middleware)
+
+	var specificationModification = `
+{
+	"secrets": {
+		"secret/myservice/external-account-key": {
+			"type": "versioned",
+			"current": "YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXowMTIzNDU=",
+			"previous": "aHVudGVyMg=="
+		}
+	},
+	"vault": {
+		"url": "vault.net",
+		"token": "17213328-36d4-11e7-8459-525400f56d04"
+	}
+}`
+	tmpFile.Write([]byte(specificationModification))
+	wg.Wait()
```

Comment:
This works when the test passes. But when there's a bug and this test doesn't pass, it could actually cause a deadlock.

Reply:
Yes. Actually I could just test it with NewStore function, no need to make an actual file change to trigger this.